### ISSUE TITLE
How to prevent one caller from getting other callers observing topics

### ISSUE THREAD

**Issue by relevance-api-topics** [patcg-individual-drafts/topics#348]
Hi Topics API experts,

We are having a question about how Topics API prevents one caller from getting other callers' observing topics.

For a fetch call, one caller can pretend to be another caller as their '<topics_caller_eTLD+1>', is that right? Since the topics in Sec-Browsing-Topics header is ingested during Request, not based on a successful Response as a prerequisite from the real authorized caller's server-side.

<img width="892" alt="Screenshot 2024-10-28 at 13 04 17" src="https://github.com/user-attachments/assets/7fc3cb1d-821d-4d69-963c-593737440709">

Did we miss any key requirement for fetch call? Thanks.

**Comment by michaelkleber** [comment]
Hi "T C", can you add your name and affiliation to your GitHub profile?  Thanks.

> For a fetch call, one caller can pretend to be another caller as their '<topics_caller_eTLD+1>', is that right? Since the topics in Sec-Browsing-Topics header is ingested during Request, not based on a successful Response as a prerequisite from the real authorized caller's server-side.

I'm a little confused.

Are you asking whether caller A can cause a request to different Site B to include the Topics header?  Yes, that's right.  But only Site B would get to _see_ the Topics included in the request headers.  You said "the caller pretend to be a real authorized caller can see the topics through a way to get request's params", but that does not work: the request params are only visible to Site B that the request is sent to, not to caller A that does the sending.

> Also for iframe case, one caller can set the origin to be another caller, to use browsingTopics() within the iframe to see other callers' observing topics.

That also isn't possible: the Topics API would be called from inside the iframe, which means only the iframe owner's site would see the topics that are returned.

**Comment by relevance-api-topics** [comment]
Hi @michaelkleber, thanks for the information, that's exactly the part we are trying to confirm:

1) using fetch call, caller A cannot know caller B's topics.
2) using iframe, caller A cannot know caller B's topics.

How do you determine if it is caller A or caller B?

for "1) using fetch call, caller A cannot know caller B's topics", if caller A sends a request to caller B's endpoint, do you check if the request is successful then fill the topic header into the fetch call? how do you determine the caller, is that from the fetch call's url.

for "2) using iframe, caller A cannot know caller B's topics", would caller A be able to create a iframe and set the src/target to be caller B's endpoint, how would you determine the iframe owner?

Thanks!

**Comment by michaelkleber** [comment]
When a fetch request is sent with the topics HTTP header, the party whose topics are included is the party whose server the HTTP request is sent to.  It doesn't matter who called the API; the only thing that matters is the destination.  I'm not sure what you mean about checking whether the request is "successful", though.

When an iframe is created, the thing that matters is the src of the iframe.  If the Topics JS API gets called from inside that iframe, the iframe's src is what determines which topics are returned.

In both cases, that means that the party which gets to see the topics is the party whose topics are being shown — which is just what we would want to be the case.

**Comment by relevance-api-topics** [comment]
Hi @michaelkleber thanks for the detailed information!

As callerA, if I call document.browsingTopics() inside a iframe with src towards callerB's server, will I get callerB's topics from client-side?

**Comment by michaelkleber** [comment]
What do you mean by "As callerA, if I call document.browsingTopics() inside a iframe with src towards callerB's server"?  How are you callerA doing anything inside callerB's iframe?  Is callerB's HTML in the iframe loading a script of yours using a third-party tag `<script src="callerA.com/foo.js">` or something?

Any JS running inside of callerB's iframe will get callerB's topics.  So if your JS is somehow running inside callerB's iframe, it would get the same results as callerA's JS or anyone else's JS running inside that same iframe: the topics visible to callerB, whose iframe it is.

**Comment by relevance-api-topics** [comment]
Hi @michaelkleber apologize for my not clear question, what I am thinking is for Chrome to determine the caller context, are you just checking the iframe's src? what if I create a iframe inside <script src="callerA.com/foo.js"> as you noted, but set src to be Google.com (random example as an instance)? can I get Google.com's topics? thanks.

**Comment by michaelkleber** [comment]
For the JS API, the caller is just determined by the iframe's src.

I still don't think you can get google.com's topics, though.  If you create an iframe with src="callerA.com", and then inside that you create another (nested) iframe with src="google.com", then the inner iframe's call to the API would return google's topics.  But those would not be visible to the outer iframe... unless the code served by google inside that inner iframe called the topics API and, I don't know, sent a postMessage to its window.parent to tell callerA's code what it saw.

**Comment by relevance-api-topics** [comment]
Hi @michaelkleber thanks for your detailed information! the question is exactly what we are thinking about 'pretending'...

Not only for iframe, if I send the arbitrary fetch call towards topics.google.com, the call may be failed, but the topics for google.com is filled within the request, correct? either by a network package wrapper or from hacky client-side code, no matter whether the fetch call is successfully or not, we could try to see google.com's topics, is this a valid thinking? thanks.

**Comment by michaelkleber** [comment]
No, I don't believe you will be able to see google.com's topics in this way either.

**Comment by relevance-api-topics** [comment]
I am not a Hacker but technically just would like to know how Topics API determine which caller it is to avoid our usage of Topics API is unsafe, if it based on src of the iframe, or the server host of the fetch call, how to avoid someone pretend to do the call. Thanks.

**Comment by relevance-api-topics** [comment]
Thanks @michaelkleber for all the details provided!
